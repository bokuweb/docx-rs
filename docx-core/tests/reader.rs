use insta::assert_json_snapshot;

use docx_rs::*;
use std::fs::*;
use std::io::{Read, Write};

#[test]
pub fn read_hello() {
    let mut file = File::open("../fixtures/hello_world/hello_world.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_hello", &json);

    let path = std::path::Path::new("./tests/output/hello.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_numbering() {
    let mut file = File::open("../fixtures/numbering/numbering.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_numbering", &json);

    let path = std::path::Path::new("./tests/output/numbering.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_decoration() {
    let mut file = File::open("../fixtures/decoration/decoration.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_decoration", &json);

    let path = std::path::Path::new("./tests/output/decoration.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_highlight_and_underline() {
    let mut file =
        File::open("../fixtures/highlight_and_underline/highlight_and_underline.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_highlight_and_underline", &json);

    let path = std::path::Path::new("./tests/output/highlight_and_underline.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_history() {
    let mut file = File::open("../fixtures/history_libre_office/history.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_history", &json);

    let path = std::path::Path::new("./tests/output/history.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_indent_word_online() {
    let mut file = File::open("../fixtures/indent_word_online/indent.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_indent_word_online", &json);

    let path = std::path::Path::new("./tests/output/indent_word_online.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_tab_and_break() {
    let mut file = File::open("../fixtures/tab_and_break/tab_and_break.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_tab_and_break", &json);

    let path = std::path::Path::new("./tests/output/tab_and_break.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_table_docx() {
    let mut file = File::open("../fixtures/table_docx/table.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_table_docx", &json);

    let path = std::path::Path::new("./tests/output/table_docx.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_table_merged_libre_office() {
    let mut file = File::open("../fixtures/table_merged_libre_office/table_merged.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_table_merged_libre_office", &json);

    let path = std::path::Path::new("./tests/output/table_merged_libre_office.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_bom() {
    let mut file = File::open("../fixtures/bom/bom.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_bom", &json);

    let path = std::path::Path::new("./tests/output/bom.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_bookmark() {
    let mut file = File::open("../fixtures/bookmark/bookmark.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_bookmark", &json);

    let path = std::path::Path::new("./tests/output/bookmark.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_insert_table() {
    let mut file = File::open("../fixtures/insert_table/insert_table.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_insert_table", &json);

    let path = std::path::Path::new("./tests/output/insert_table.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_textbox() {
    let mut file = File::open("../fixtures/textbox/textbox.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_textbox", &json);

    let path = std::path::Path::new("./tests/output/textbox.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_from_doc() {
    let mut file = File::open("../fixtures/from_doc/from_doc.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_from_doc", &json);

    let path = std::path::Path::new("./tests/output/from_doc.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_lvl_override() {
    let mut file = File::open("../fixtures/lvl_override/override.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_lvl_override", &json);

    let path = std::path::Path::new("./tests/output/lvl_override.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_comment() {
    let mut file = File::open("../fixtures/comment/comment.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_comment", &json);

    let path = std::path::Path::new("./tests/output/comment.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_extended_comment() {
    let mut file = File::open("../fixtures/extended_comments/extended_comments.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_extended_comments", &json);

    let path = std::path::Path::new("./tests/output/extended_comments.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_line_spacing() {
    let mut file = File::open("../fixtures/line_spacing/line_spacing.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("line_spacing", &json);

    let path = std::path::Path::new("./tests/output/line_spacing.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}
