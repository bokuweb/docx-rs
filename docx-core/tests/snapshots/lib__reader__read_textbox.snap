---
source: docx-core/tests/reader.rs
expression: "&json"
---
"{\n  \"contentType\": {\n    \"types\": {\n      \"/_rels/.rels\": \"application/vnd.openxmlformats-package.relationships+xml\",\n      \"/docProps/app.xml\": \"application/vnd.openxmlformats-officedocument.extended-properties+xml\",\n      \"/docProps/core.xml\": \"application/vnd.openxmlformats-package.core-properties+xml\",\n      \"/word/_rels/document.xml.rels\": \"application/vnd.openxmlformats-package.relationships+xml\",\n      \"/word/comments.xml\": \"application/vnd.openxmlformats-officedocument.wordprocessingml.comments+xml\",\n      \"/word/commentsExtended.xml\": \"application/vnd.openxmlformats-officedocument.wordprocessingml.commentsExtended+xml\",\n      \"/word/document.xml\": \"application/vnd.openxmlformats-officedocument.wordprocessingml.document.main+xml\",\n      \"/word/fontTable.xml\": \"application/vnd.openxmlformats-officedocument.wordprocessingml.fontTable+xml\",\n      \"/word/header1.xml\": \"application/vnd.openxmlformats-officedocument.wordprocessingml.header+xml\",\n      \"/word/numbering.xml\": \"application/vnd.openxmlformats-officedocument.wordprocessingml.numbering+xml\",\n      \"/word/settings.xml\": \"application/vnd.openxmlformats-officedocument.wordprocessingml.settings+xml\",\n      \"/word/styles.xml\": \"application/vnd.openxmlformats-officedocument.wordprocessingml.styles+xml\"\n    }\n  },\n  \"rels\": {\n    \"rels\": [\n      [\n        \"http://schemas.openxmlformats.org/package/2006/relationships/metadata/core-properties\",\n        \"rId1\",\n        \"docProps/core.xml\"\n      ],\n      [\n        \"http://schemas.openxmlformats.org/officeDocument/2006/relationships/extended-properties\",\n        \"rId2\",\n        \"docProps/app.xml\"\n      ],\n      [\n        \"http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument\",\n        \"rId3\",\n        \"word/document.xml\"\n      ]\n    ]\n  },\n  \"documentRels\": {\n    \"hasComments\": false,\n    \"hasNumberings\": false,\n    \"imageIds\": []\n  },\n  \"docProps\": {\n    \"app\": {},\n    \"core\": {\n      \"config\": {\n        \"created\": null,\n        \"creator\": null,\n        \"description\": null,\n        \"language\": null,\n        \"lastModifiedBy\": null,\n        \"modified\": null,\n        \"revision\": null,\n        \"subject\": null,\n        \"title\": null\n      }\n    }\n  },\n  \"styles\": {\n    \"docDefaults\": {\n      \"runPropertyDefault\": {\n        \"runProperty\": {\n          \"sz\": null,\n          \"szCs\": null,\n          \"color\": null,\n          \"highlight\": null,\n          \"underline\": null,\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": null,\n          \"italicCs\": null,\n          \"vanish\": null,\n          \"fonts\": null\n        }\n      }\n    },\n    \"styles\": [\n      {\n        \"styleId\": \"a\",\n        \"name\": \"Normal\",\n        \"styleType\": \"paragraph\",\n        \"runProperty\": {\n          \"sz\": null,\n          \"szCs\": null,\n          \"color\": null,\n          \"highlight\": null,\n          \"underline\": null,\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": null,\n          \"italicCs\": null,\n          \"vanish\": null,\n          \"fonts\": null\n        },\n        \"paragraphProperty\": {\n          \"runProperty\": {\n            \"sz\": null,\n            \"szCs\": null,\n            \"color\": null,\n            \"highlight\": null,\n            \"underline\": null,\n            \"bold\": null,\n            \"boldCs\": null,\n            \"italic\": null,\n            \"italicCs\": null,\n            \"vanish\": null,\n            \"fonts\": null\n          },\n          \"style\": null,\n          \"numberingProperty\": null,\n          \"alignment\": \"both\",\n          \"indent\": null\n        }\n      },\n      {\n        \"styleId\": \"a0\",\n        \"name\": \"Default Paragraph Font\",\n        \"styleType\": \"character\",\n        \"runProperty\": {\n          \"sz\": null,\n          \"szCs\": null,\n          \"color\": null,\n          \"highlight\": null,\n          \"underline\": null,\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": null,\n          \"italicCs\": null,\n          \"vanish\": null,\n          \"fonts\": null\n        },\n        \"paragraphProperty\": {\n          \"runProperty\": {\n            \"sz\": null,\n            \"szCs\": null,\n            \"color\": null,\n            \"highlight\": null,\n            \"underline\": null,\n            \"bold\": null,\n            \"boldCs\": null,\n            \"italic\": null,\n            \"italicCs\": null,\n            \"vanish\": null,\n            \"fonts\": null\n          },\n          \"style\": null,\n          \"numberingProperty\": null,\n          \"alignment\": null,\n          \"indent\": null\n        }\n      },\n      {\n        \"styleId\": \"a1\",\n        \"name\": \"Normal Table\",\n        \"styleType\": \"table\",\n        \"runProperty\": {\n          \"sz\": null,\n          \"szCs\": null,\n          \"color\": null,\n          \"highlight\": null,\n          \"underline\": null,\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": null,\n          \"italicCs\": null,\n          \"vanish\": null,\n          \"fonts\": null\n        },\n        \"paragraphProperty\": {\n          \"runProperty\": {\n            \"sz\": null,\n            \"szCs\": null,\n            \"color\": null,\n            \"highlight\": null,\n            \"underline\": null,\n            \"bold\": null,\n            \"boldCs\": null,\n            \"italic\": null,\n            \"italicCs\": null,\n            \"vanish\": null,\n            \"fonts\": null\n          },\n          \"style\": null,\n          \"numberingProperty\": null,\n          \"alignment\": null,\n          \"indent\": null\n        }\n      },\n      {\n        \"styleId\": \"a2\",\n        \"name\": \"No List\",\n        \"styleType\": \"numbering\",\n        \"runProperty\": {\n          \"sz\": null,\n          \"szCs\": null,\n          \"color\": null,\n          \"highlight\": null,\n          \"underline\": null,\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": null,\n          \"italicCs\": null,\n          \"vanish\": null,\n          \"fonts\": null\n        },\n        \"paragraphProperty\": {\n          \"runProperty\": {\n            \"sz\": null,\n            \"szCs\": null,\n            \"color\": null,\n            \"highlight\": null,\n            \"underline\": null,\n            \"bold\": null,\n            \"boldCs\": null,\n            \"italic\": null,\n            \"italicCs\": null,\n            \"vanish\": null,\n            \"fonts\": null\n          },\n          \"style\": null,\n          \"numberingProperty\": null,\n          \"alignment\": null,\n          \"indent\": null\n        }\n      }\n    ]\n  },\n  \"document\": {\n    \"children\": [\n      {\n        \"type\": \"paragraph\",\n        \"data\": {\n          \"id\": \"0000006c\",\n          \"children\": [\n            {\n              \"type\": \"run\",\n              \"data\": {\n                \"runProperty\": {\n                  \"sz\": null,\n                  \"szCs\": null,\n                  \"color\": null,\n                  \"highlight\": null,\n                  \"underline\": null,\n                  \"bold\": null,\n                  \"boldCs\": null,\n                  \"italic\": null,\n                  \"italicCs\": null,\n                  \"vanish\": null,\n                  \"fonts\": null\n                },\n                \"children\": [\n                  {\n                    \"type\": \"drawing\",\n                    \"data\": {\n                      \"position_type\": {\n                        \"Inline\": {\n                          \"dist_t\": 0,\n                          \"dist_b\": 0,\n                          \"dist_l\": 0,\n                          \"dist_r\": 0\n                        }\n                      },\n                      \"position_h\": {\n                        \"Offset\": 0\n                      },\n                      \"position_v\": {\n                        \"Offset\": 0\n                      },\n                      \"data\": null,\n                      \"children\": [\n                        {\n                          \"type\": \"anchor\",\n                          \"data\": {\n                            \"children\": [\n                              {\n                                \"children\": [\n                                  {\n                                    \"dataType\": \"wpShape\",\n                                    \"children\": [\n                                      {\n                                        \"type\": \"shape\",\n                                        \"data\": {\n                                          \"children\": [\n                                            {\n                                              \"type\": \"textbox\",\n                                              \"data\": {\n                                                \"children\": [\n                                                  {\n                                                    \"children\": [\n                                                      {\n                                                        \"type\": \"paragraph\",\n                                                        \"data\": {\n                                                          \"id\": \"0000006d\",\n                                                          \"children\": [\n                                                            {\n                                                              \"type\": \"run\",\n                                                              \"data\": {\n                                                                \"runProperty\": {\n                                                                  \"sz\": null,\n                                                                  \"szCs\": null,\n                                                                  \"color\": null,\n                                                                  \"highlight\": null,\n                                                                  \"underline\": null,\n                                                                  \"bold\": null,\n                                                                  \"boldCs\": null,\n                                                                  \"italic\": null,\n                                                                  \"italicCs\": null,\n                                                                  \"vanish\": null,\n                                                                  \"fonts\": null\n                                                                },\n                                                                \"children\": [\n                                                                  {\n                                                                    \"type\": \"text\",\n                                                                    \"data\": {\n                                                                      \"preserveSpace\": true,\n                                                                      \"text\": \"H\"\n                                                                    }\n                                                                  }\n                                                                ]\n                                                              }\n                                                            },\n                                                            {\n                                                              \"type\": \"run\",\n                                                              \"data\": {\n                                                                \"runProperty\": {\n                                                                  \"sz\": null,\n                                                                  \"szCs\": null,\n                                                                  \"color\": null,\n                                                                  \"highlight\": null,\n                                                                  \"underline\": null,\n                                                                  \"bold\": null,\n                                                                  \"boldCs\": null,\n                                                                  \"italic\": null,\n                                                                  \"italicCs\": null,\n                                                                  \"vanish\": null,\n                                                                  \"fonts\": null\n                                                                },\n                                                                \"children\": [\n                                                                  {\n                                                                    \"type\": \"text\",\n                                                                    \"data\": {\n                                                                      \"preserveSpace\": true,\n                                                                      \"text\": \"ello\"\n                                                                    }\n                                                                  }\n                                                                ]\n                                                              }\n                                                            }\n                                                          ],\n                                                          \"property\": {\n                                                            \"runProperty\": {\n                                                              \"sz\": null,\n                                                              \"szCs\": null,\n                                                              \"color\": null,\n                                                              \"highlight\": null,\n                                                              \"underline\": null,\n                                                              \"bold\": null,\n                                                              \"boldCs\": null,\n                                                              \"italic\": null,\n                                                              \"italicCs\": null,\n                                                              \"vanish\": null,\n                                                              \"fonts\": null\n                                                            },\n                                                            \"style\": null,\n                                                            \"numberingProperty\": null,\n                                                            \"alignment\": null,\n                                                            \"indent\": null\n                                                          },\n                                                          \"hasNumbering\": false\n                                                        }\n                                                      }\n                                                    ],\n                                                    \"has_numbering\": false\n                                                  }\n                                                ],\n                                                \"hasNumbering\": false\n                                              }\n                                            }\n                                          ]\n                                        }\n                                      }\n                                    ]\n                                  }\n                                ]\n                              }\n                            ]\n                          }\n                        }\n                      ]\n                    }\n                  }\n                ]\n              }\n            }\n          ],\n          \"property\": {\n            \"runProperty\": {\n              \"sz\": null,\n              \"szCs\": null,\n              \"color\": null,\n              \"highlight\": null,\n              \"underline\": null,\n              \"bold\": null,\n              \"boldCs\": null,\n              \"italic\": null,\n              \"italicCs\": null,\n              \"vanish\": null,\n              \"fonts\": null\n            },\n            \"style\": null,\n            \"numberingProperty\": null,\n            \"alignment\": null,\n            \"indent\": null\n          },\n          \"hasNumbering\": false\n        }\n      }\n    ],\n    \"sectionProperty\": {\n      \"pageSize\": {\n        \"w\": 11906,\n        \"h\": 16838\n      },\n      \"pageMargin\": {\n        \"top\": 1985,\n        \"left\": 1701,\n        \"bottom\": 1701,\n        \"right\": 1701,\n        \"header\": 851,\n        \"footer\": 992,\n        \"gutter\": 0\n      },\n      \"columns\": 425,\n      \"documentGrid\": 360,\n      \"headerReference\": {\n        \"headerType\": \"default\",\n        \"id\": \"rId4\"\n      },\n      \"sectionType\": null\n    },\n    \"hasNumbering\": false\n  },\n  \"comments\": {\n    \"comments\": []\n  },\n  \"numberings\": {\n    \"abstractNums\": [],\n    \"numberings\": []\n  },\n  \"settings\": {\n    \"defaultTabStop\": 709,\n    \"zoom\": 100,\n    \"docId\": null\n  },\n  \"fontTable\": {},\n  \"media\": [],\n  \"header\": {\n    \"children\": []\n  },\n  \"commentsExtended\": {\n    \"children\": []\n  }\n}"
